certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+}
+
+
+// 10. removeMember() - removing a member does not add/remove any other member
+// Status: Pass
+// https://prover.certora.com/output/80942/0711dedd9b0847bd9e5ef3741e6f0017/?anonymousKey=28bc4551b778792663a295c140e8b0c60369ae38
+rule removeMemberADoesNotModifyMemberB() {
+    env e; env e2;
+    address userA; uint256 quorum; address userB;
+    bool isUserAMemberBefore = getIsMember(e,userA);
+    bool isUserBMemberBefore = getIsMember(e,userB);
+    uint256 index1bUserABefore = get_memberIndices1b(e,userA);
+    uint256 index1bUserBBefore = get_memberIndices1b(e,userB);
+
+    uint256 lengthOf_memberAddressesBefore; uint256 lengthOf_memberStatesBefore;
+    lengthOf_memberAddressesBefore, lengthOf_memberStatesBefore = getLengthOfArrays(e);  // using helper
+
+    if (!isUserBMemberBefore) {
+        address user = get_memberAddresses(e,lengthOf_memberAddressesBefore - 1);
+        require user != userB;      // safe assumption, if userB is not a member that must hold too (verify?)
+    }
+
+    require lengthOf_memberAddressesBefore == lengthOf_memberStatesBefore;  // those arrays represent the same members
+
+    removeMember(e, userA, quorum);                         // call removeMember to remove userA
+
+    bool isUserAMemberAfter = getIsMember(e2,userA);
+    bool isUserBMemberAfter = getIsMember(e2,userB);
+    uint256 index1bUserAAfter = get_memberIndices1b(e,userA);
+    uint256 index1bUserBAfter = get_memberIndices1b(e,userB);
+
+    require userA != userB;                                 // userA and userB are different
+
+    assert isUserBMemberBefore => isUserBMemberAfter;       // adding userA does not remove userB
+    assert !isUserBMemberBefore => !isUserBMemberAfter;     // adding userA does not add userB
+}
+
+
+// 11. removeMember() - removing a member decreases the total members by 1
+// Status: Pass
+// https://prover.certora.com/output/80942/3f82ae3e6cd044578656753520da3cec/?anonymousKey=0ed4f43d9f0ea4d0c923e368f8704f635d2559c0
+rule removeMemberCorrectness() {
+    env e; env e2;
+
+    uint256 lengthOf_memberAddressesBefore; uint256 lengthOf_memberStatesBefore;
+    lengthOf_memberAddressesBefore, lengthOf_memberStatesBefore = getLengthOfArrays(e);  // using helper
+
+    require lengthOf_memberAddressesBefore == lengthOf_memberStatesBefore;  // those arrays represent the same members
+
+    address userA; uint256 quorum;
+    bool isUserAMemberBefore = getIsMember(e,userA);
+
+    removeMember(e, userA, quorum);                            // call removeMember to remove user userA
+
+    uint256 lengthOf_memberAddressesAfter; uint256 lengthOf_memberStatesAfter;
+    lengthOf_memberAddressesAfter, lengthOf_memberStatesAfter = getLengthOfArrays(e);  // using helper
+    bool isUserAMemberAfter = getIsMember(e2,userA);
+
+    assert isUserAMemberBefore;                            // cannot remove userA if he was not a member before
```

Comment:
it seems the same as with addMember(). You call removeMember() without @withrevert, so this asserts are guaranteed